### PR TITLE
PRT-1046 Adding error handling and timeout parsing for better UX for the user.

### PR DESCRIPTION
**Pull request by ranlavanet**

### CODE REVIEW

File: protocol/rpcconsumer/relay_errors.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		}
+		currentResult := relayError.ProviderInfo.ProviderQoSExcellenceSummery.MulInt(relayError.ProviderInfo.ProviderStake.Amount)
+		if currentResult.GTE(bestResult) { // 0 or 1 here are valid replacements, so even 0 scores will return the error value
+			bestResult.Set(currentResult)
+			bestIndex = idx
+		}
+	}
+
+	errorKey, errorCount := r.findLargestValueInMap(errorMap)
+	if errorCount > (len(r.relayErrors) / 2) {
```
Comment (omerlavanet) [inline protocol/rpcconsumer/relay_errors.go:45]: greater equal

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: protocol/lavasession/consumer_session_manager.go [hunk 2]
```diff
@@ -401,7 +402,7 @@ func (csm *ConsumerSessionManager) GetSessions(ctx context.Context, cuNeededForS
 			}
 
 			// If we successfully got a consumerSession we can apply the current CU to the consumerSessionWithProvider.UsedComputeUnits
-			err = consumerSessionsWithProvider.addUsedComputeUnits(cuNeededForSession, virtualEpoch)
+			providerStakeSize, err := consumerSessionsWithProvider.addUsedComputeUnitsAndReturnProviderStakeSize(cuNeededForSession, virtualEpoch)
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply protocol/lavasession/consumer_session_manager.go]: done 91ae92e541b31c2c6b47689659f2b69268691876

File: protocol/lavasession/consumer_types.go [hunk 3]
```diff
@@ -251,15 +254,15 @@ func (cswp *ConsumerSessionsWithProvider) validateComputeUnits(cu uint64, virtua
...
 	cswp.Lock.Lock()
 	defer cswp.Lock.Unlock()
 	// add additional CU for virtual epochs
 	if (cswp.UsedComputeUnits + cu) > cswp.MaxComputeUnits*(virtualEpoch+1) {
-		return MaxComputeUnitsExceededError
+		return cswp.StakeSize, MaxComputeUnitsExceededError
 	}
 	cswp.UsedComputeUnits += cu
-	return nil
+	return cswp.StakeSize, nil
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply protocol/lavasession/consumer_types.go]: needs to be locked. fixed here: 9b7fb7060800aa00ed5565f8e0f64dfbc56ddc7c

File: protocol/lavasession/consumer_session_manager.go [hunk 4]
```diff
@@ -426,12 +427,26 @@ func (csm *ConsumerSessionManager) GetSessions(ctx context.Context, cuNeededForS
...
 					Epoch:             sessionEpoch,
 					ReportedProviders: reportedProviders,
 				}
+				// adding qos summery for error parsing.
+				// consumer session is locked here so its ok to read the qos report.
+				if consumerSession.QoSInfo.LastExcellenceQoSReport != nil {
+					qosComputed, errComputing := consumerSession.QoSInfo.LastExcellenceQoSReport.ComputeQoS()
+					if errComputing != nil { // if we failed to compute the qos will be 0 so this provider wont be picked to return the error in case we get it
+						utils.LavaFormatError("Failed computing QoS used for error parsing", errComputing, utils.LogAttr("Report", consumerSession.QoSInfo.LastExcellenceQoSReport))
+						sessionInfo.QoSSummeryResult = github_com_cosmos_cosmos_sdk_types.ZeroDec()
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply protocol/lavasession/consumer_session_manager.go]: Solved here: e1ca02104bfee61ef320d3b3c53c8f828dc682a3

File: protocol/rpcconsumer/relay_errors.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	github_com_cosmos_cosmos_sdk_types "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/protocol/common"
+	"github.com/lavanet/lava/utils"
+)
+
+type RelayErrors struct {
+	relayErrors []RelayError
+}
+
+func (r *RelayErrors) findLargestValueInMap(myMap map[string]int) (string, int) {
```
[the remark this answers is not in the document]
Comment (ranlavanet) [reply protocol/rpcconsumer/relay_errors.go]: fixed here: 16e01cd8e598dc6a59a5b4c47ac0c9a729512d07

File: protocol/rpcconsumer/relay_errors.go (hunk 1 above)
Comment (ranlavanet) [reply protocol/rpcconsumer/relay_errors.go]: fixed here: 16e01cd8e598dc6a59a5b4c47ac0c9a729512d07

**Review by omerlavanet — APPROVED** [review]

**Pull request merged**